sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala

```diff
@@ -62,111 +100,250 @@ case class SortMergeJoin(
+                val joinedRow = smartJoinRow(streamedElement, bufferedNullRow)
+                fetchStreamed()
+                joinedRow
+              } else {
+                val joinedRow = smartJoinRow(streamedNullRow, bufferedElement)
+                fetchBuffered()
+                joinedRow
               }
+            } else {
+              // we are using the buffered right rows and run down left iterator
+              val joinedRow = if (streamNullGenerated) {
+                val ret = smartJoinRow(streamedNullRow, bufferedMatches(bufferedPosition))
+                bufferedPosition += 1
+                ret
+              } else {
+                if (bufferedPosition == bufferedMatches.size && !bufferNullGenerated) {
+                  val ret = smartJoinRow(streamedElement, bufferedNullRow)
+                  bufferNullGenerated = true
+                  ret
+                } else {
+                  val ret = smartJoinRow(streamedElement, bufferedMatches(bufferedPosition))
+                  bufferedPosition += 1
+                  ret
+                }
+              }
+              found = false
+              joinedRow
             }
-            joinedRow
           } else {
             // no more result
             throw new NoSuchElementException
           }
         }
 
-        private def fetchLeft() = {
-          if (leftIter.hasNext) {
-            leftElement = leftIter.next()
-            leftKey = leftKeyGenerator(leftElement)
+        private def smartJoinRow(streamedRow: InternalRow, bufferedRow: InternalRow): InternalRow =
+          joinType match {
+            case RightOuter => joinRow(bufferedRow, streamedRow)
+            case _ => joinRow(streamedRow, bufferedRow)
+          }
+
+        private def fetchStreamed(): Unit = {
+          if (streamedIter.hasNext) {
+            streamedElement = streamedIter.next()
+            streamedKey = streamedKeyGenerator(streamedElement)
           } else {
-            leftElement = null
+            streamedElement = null
           }
         }
 
-        private def fetchRight() = {
-          if (rightIter.hasNext) {
-            rightElement = rightIter.next()
-            rightKey = rightKeyGenerator(rightElement)
+        private def fetchBuffered(): Unit = {
+          if (bufferedIter.hasNext) {
+            bufferedElement = bufferedIter.next()
+            bufferedKey = bufferedKeyGenerator(bufferedElement)
           } else {
-            rightElement = null
+            bufferedElement = null
           }
         }
 
         private def initialize() = {
-          fetchLeft()
-          fetchRight()
+          fetchStreamed()
+          fetchBuffered()
         }
 
         /**
-         * Searches the right iterator for the next rows that have matches in left side, and store
-         * them in a buffer.
+         * Searches the right iterator for the next rows that have matches in left side (only check
```

Comment:
Is it confusing because right and left can be both streamed or buffered here? Do we need to use streamed and buffered in the comments as well?